### DISCUSSION TITLE
Proposal for API Consolidation in Modules

### DISCUSSION THREAD

**Discussion by yairgott** [valkey-io/valkey#371] [General]
Modules provide a great way to enhance ValKey functionality and to grow our community.  One challenge with a community-driven approach for modules is the risk of contributors evolving their APIs in different directions. This divergence may lead to difficulties in usability, sustaining clients being up to date, and more.

To benefit ValKey and our users, the TSC committee may establish standardization around APIs and general clear module qualification criterias. Ideally, this standardization would be driven not directly by the TSC, but by a subcommittee with domain expertise and credibility in a particular domain. Meeting the qualification means that the module is endorsed by ValKey. To clarify, my proposal is saying that it is perfectly fine to have more than a single qualified module per domain, like json, vector search, etc. At the end of the day, we let the user decide and let the modules organically evolve over time.

Happy to get your thoughts.

**Comment by madolson** [comment]
@yairgott Just to make sure I understand what you're saying, you're not talking about the API between the module and Valkey, you're talking about the API between the end user and the module?

Having working groups per module seems like a good idea. I would like to see the TSC oversee decisions, but we already have the primitives in place in the governance draft to allow them to delegate decisions.

**Reply by yairgott** [reply]
@madolson , thanks for the reply and yes, I'm specifically talking about the API between the end user and the module.

I agree that especially at the beginning until trust is formed and a subcommittee establishes credibility, it's important that TSC oversee decisions and more closely involved. Long term, the aspiration is for the subcommittee to become independent.

**Reply by madolson** [reply]
I don't see a future in which the subcommittee's become truly independent, since I think the TSC should remain as the final say for the direction of the project. A subcommittee can always go do it's own thing and start its own project if they think the community is doing the wrong thing.
